### ISSUE TITLE
Keyboard events not passed to terminal UI apps

### ISSUE THREAD

**Issue by nojus297** [microsoft/vscode#90838]
**VScode:**
Version: 1.41.1
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-18T15:04:31.999Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Linux x64 4.19.0-6-amd64

**Description:**
Some keyboard presses are not passed to a terminal UI app, running inside integrated terminal. It seems like they are handled and consumed by vscode itself, even though they could be used by the terminal UI app.

**Steps to Reproduce:**
1.  Open integrated terminal
1.  Launch `nano` editor containing multi-line text with cursor in the first line 
1.  Press `ctrl-shif-arrowDown`
1.  Add `{
        "key": "ctrl+shift+down",
        "command": "-workbench.action.terminal.scrollDown",
        "when": "terminalFocus"
    },` to `keybindings.json`.
1.  Press `ctrl-shif-arrowDown`again in the `nano` editor
1.  Repeat steps 2-3 in external terminal like `gnome-terminal` or `xfce4-terminal`.

**Observed behavior**
Step 3 - nothing happens. Step 5 and 6 - text is selected up to the end of the `nano`'s buffer.

**Expected behavior:**
The shortcut should work without disabling integrated terminal scrolling in keybindings. Maybe vscode shouldn't consume key events that could be used by terminal UI apps. This is the way it works in external terminals (`ctrl-shif-arrowDown` - scrolls down in normal mode, but is usable for launched terminal UI apps).  
In vscode `ctrl-shif-arrowDown` is not even delivered to the terminal UI app.

Does this issue occur when all extensions are disabled?: Yes

**Comment by IllusionMH** [comment]
`terminal.integrated.commandsToSkipShell` controls shortcuts to which commands should be intercepted by VS Code and has [long default list](https://github.com/microsoft/vscode/blob/2735ee99085cac449568b4702639134c8eb3b85c/src/vs/workbench/contrib/terminal/browser/terminalInstance.ts#L50-L167).

You can try to remove actions from list so they will be passes to terminal.
Sine you've mentioned `workbench.action.terminal.scrollDown` it would be
```json
"terminal.integrated.commandsToSkipShell": [
    "-workbench.action.terminal.scrollDown"
]
```
However you'll need to check and add to list commands that match your shortcuts on Linux.

**Comment by Nojus297** [comment]
Thanks for the quick reply, however, your suggested fix does the same as removing keybinding - removes scrolling functionality in terminal using `ctrl-shif-arrowDown`

**Comment by IllusionMH** [comment]
Because now Ctrl+Shift+ArrowDown is ignored by VSCode and instead will be sent to terminal This way it can be processed by nano. Didn’t you want to achieve this?

**Comment by Nojus297** [comment]
Yes, but I also want it to act as a scrolling keybinding in terminal (scrolling history). Basically, I want both functions, not just a single one.

**Comment by IllusionMH** [comment]
IMHO it's either-or.
How VS Code should determine when to intercept or when to pass shortcut to terminal?
Probably it would require to support big map (which won't be full or outdated) of shortcuts for different process names.
However it still won't be useful - what if nano has some action on e.g. Ctrl+P but I want Ctrl+P to be processed by VS Code? What if it's another way?

/cc @Tyriar

**Comment by Nojus297** [comment]
I don't know either how vscode should or shouldn't handle this, but yet external terminal emulators are able to provide both functions. Maybe when a terminal UI app is running inside the terminal, it should pass all the navigation key events to the app?

**Comment by Tyriar** [comment]
> Maybe when a terminal UI app is running inside the terminal, it should pass all the navigation key events to the app?

We haven't yet gone this route but we could. There is no scrolling when you're in the alt buffer (eg. when in nano), so we could detect this and pass on if so. We would probably need another context key though. Something like this:

```
{
  "key": "ctrl+shift+down",
  "command": "-workbench.action.terminal.scrollDown",
  "when": "terminalFocus && terminalAltBufferActive"
}
```

This is blocked on API from upstream lib: https://github.com/xtermjs/xterm.js/issues/2694

**Comment by Nojus297** [comment]
Yeah, this should work.
Maybe this could be set by default, because  IMHO there isn't any point to consume navigation key events when terminal buffer is not even scrollable.

**Comment by lilyball** [comment]
I strongly support the idea of allowing keyboard shortcuts to go to the terminal when the alt buffer is active. In addition, once this is done, the Home/End keys should be flipped around so they'll scroll by default, but go to the terminal when the alt buffer is active (today Home/End aren't intercepted by default and so go to the terminal, and Cmd-Home/End scrolls the terminal).

There should also be default bindings to scroll anyway while the alt buffer is active. macOS Terminal.app uses the Shift modifier for this, which seems like a good choice as AFAIK there's no terminal representation for "Shift-Home" or "Shift-PageDn" anyway. VSCode could keep the current Cmd-Home/End for this but that feels like an odd choice from the beginning, and it doesn't currently have a Cmd-PageUp/Down binding in the terminal anyway.

Also to be clear, Terminal.app doesn't just use Shift to mean "scroll anyway", it uses it to mean "invert the behavior". So when the alt buffer is active, PageDown will go to the terminal and Shift-PageDown will scroll. But when the alt buffer isn't active, PageDown will scroll and Shift-PageDown will go to the terminal. And this really feels like the right behavior to use.

**Comment by Tyriar** [comment]
Pushed the change to disable these while the alt buffer is active.

> Also to be clear, Terminal.app doesn't just use Shift to mean "scroll anyway", it uses it to mean "invert the behavior". So when the alt buffer is active, PageDown will go to the terminal and Shift-PageDown will scroll. But when the alt buffer isn't active, PageDown will scroll and Shift-PageDown will go to the terminal. And this really feels like the right behavior to use.

@lilyball this would be a much more impactful change so I kept it out of this fix, but we can discuss in a separate issue if you would like.
